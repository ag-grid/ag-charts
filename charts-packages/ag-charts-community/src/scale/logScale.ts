import { ContinuousScale } from "./continuousScale";
import ticks from "../util/ticks";
import { format } from "../util/numberFormat";

const identity = (x: any) => x;
export class LogScale extends ContinuousScale {
    readonly type = 'log';

    _domain = [1, 10];

    baseLog = identity; // takes a log with base `base` of `x`
    basePow = identity; // raises `base` to the power of `x`

    protected setDomain(values: any[]) {
        let df = values[0];
        let dl = values[values.length - 1];

        if (df === 0 || dl === 0 || df < 0 && dl > 0 || df > 0 && dl < 0) {
            console.warn('Log scale domain should not start at, end at or cross zero.');
            if (df === 0 && dl > 0) {
                df = Number.EPSILON;
            } else if (dl === 0 && df < 0) {
                dl = -Number.EPSILON;
            } else if (df < 0 && dl > 0) {
                if (Math.abs(dl) >= Math.abs(df)) {
                    df = Number.EPSILON;
                } else {
                    dl = -Number.EPSILON;
                }
            } else if (df > 0 && dl < 0) {
                if (Math.abs(dl) >= Math.abs(df)) {
                    df = -Number.EPSILON;
                } else {
                    dl = Number.EPSILON;
                }
            }
            values = values.slice();
            values[0] = df;
            values[values.length - 1] = dl;
        }

        super.setDomain(values);
    }
    protected getDomain(): any[] {
        return super.getDomain();
    }

    _base = 10;
    set base(value: number) {
        if (this._base !== value) {
            this._base = value;
            this.rescale();
        }
    }
    get base(): number {
        return this._base;
    }

    rescale() {
        const { base } = this;
        let baseLog = LogScale.makeLogFn(base);
        let basePow = LogScale.makePowFn(base);

        if (this.domain[0] < 0) {
            baseLog = this.reflect(baseLog);
            basePow = this.reflect(basePow);
            this.transform = (x) => -Math.log(-x);
            this.untransform = (x) => -Math.exp(-x);
        } else {
            this.transform = (x) => Math.log(x);
            this.untransform = (x) => Math.exp(x);
        }

        this.baseLog = baseLog;
        this.basePow = basePow;

        super.rescale();
    }

    /**
     * For example, if `f` is `Math.log10`, we have
     *
     *     f(100) == 2
     *     f(-100) == NaN
     *     rf = reflect(f)
     *     rf(-100) == -2
     *
     * @param f
     */
    reflect(f: (x: number) => number): (x: number) => number {
        return (x: number) => -f(-x);
    }

    nice() {
        const domain = this.domain;
        let i0 = 0;
        let i1 = domain.length - 1;
        let x0 = domain[i0];
        let x1 = domain[i1];

        if (x1 < x0) {
            [i0, i1] = [i1, i0];
            [x0, x1] = [x1, x0];
        }

        // For example, for base == 10:
        // [ 50, 900] becomes [ 10, 1000 ]
        domain[i0] = this.basePow(Math.floor(this.baseLog(x0)));
        domain[i1] = this.basePow(Math.ceil(this.baseLog(x1)));

        this.domain = domain;
    }

    static pow10(x: number): number {
        return isFinite(x)
            ? +('1e' + x) // to avoid precision issues, e.g. Math.pow(10, -4) is not 0.0001
            : x < 0
                ? 0
                : x;
    }

    static makePowFn(base: number): (x: number) => number {
        if (base === 10) {
            return LogScale.pow10;
        }
        if (base === Math.E) {
            return Math.exp;
        }
        return (x: number) => Math.pow(base, x);
    }

    // Make a log function witn an arbitrary base or return a native function if exists.
    static makeLogFn(base: number) {
        if (base === Math.E) {
            return Math.log;
        }
        if (base === 10) {
            return Math.log10;
        }
        if (base === 2) {
            return Math.log2;
        }
        const logBase = Math.log(base);
        return (x: number) => Math.log(x) / logBase;
    }

    ticks(count = 10) {
        const n = count == null ? 10 : +count;
        const base = this.base;
        const domain = this.domain;
        let d0 = domain[0];
        let d1 = domain[domain.length - 1];
        const isReversed = d1 < d0;

        if (isReversed) {
            [d0, d1] = [d1, d0];
        }

        let p0 = this.baseLog(d0);
        let p1 = this.baseLog(d1);
        let z = [];

        // if `base` is an integer and delta in order of magnitudes is less than n
        if (!(base % 1) && p1 - p0 < n) {
            // For example, if n == 10, base == 10 and domain == [10^2, 10^6]
            // then p1 - p0 < n == true.
            p0 = Math.round(p0) - 1;
            p1 = Math.round(p1) + 1;
            if (d0 > 0) {
                for (; p0 < p1; ++p0) {
                    for (let k = 1, p = this.basePow(p0); k < base; ++k) {
                        let t = p * k;
                        // The `t` checks are needed because we expanded the [p0, p1] by 1 in each direction.
                        if (t < d0)
                            continue;
                        if (t > d1)
                            break;
                        z.push(t);
                    }
                }
            } else {
                for (; p0 < p1; ++p0) {
                    for (let k = base - 1, p = this.basePow(p0); k >= 1; --k) {
                        let t = p * k;
                        if (t < d0)
                            continue;
                        if (t > d1)
                            break;
                        z.push(t);
                    }
                }
            }
            if (z.length * 2 < n) {
                z = ticks(d0, d1, n);
            }
        } else {
            // For example, if n == 4, base == 10 and domain == [10^2, 10^6]
            // then p1 - p0 < n == false.
            // `ticks` return [2, 3, 4, 5, 6], then mapped to [10^2, 10^3, 10^4, 10^5, 10^6].
            z = ticks(p0, p1, Math.min(p1 - p0, n)).map(this.basePow);
        }

        return isReversed ? z.reverse() : z;
    }

    tickFormat(count: any, specifier?: string | ((x: number) => string)): (x: number) => string {
        const { base } = this;

        if (specifier == null) {
            specifier = (base === 10 ? '.0e' : ',') as any;
        }

        if (typeof specifier !== 'function') {
            specifier = format(specifier as string);
        }

        if (count === Infinity) {
            return specifier;
        }

        if (count == null) {
            count = 10;
        }

        const k = Math.max(1, base * count / this.ticks().length);

        return (d) => {
            let i = d / this.basePow(Math.round(this.baseLog(d)));
            if (i * base < base - 0.5) {
                i *= base;
            }
            return i <= k ? (specifier as (x: number) => string)(d) : '';
        };
    }
}
