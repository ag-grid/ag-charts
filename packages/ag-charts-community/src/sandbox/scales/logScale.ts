import { convertDomain } from '../util/domain.util';
import { LinearScale } from './linearScale';

export class LogScale extends LinearScale {
    readonly log: (n: number) => number;
    readonly pow: (n: number) => number;

    constructor(
        domain: number[],
        range: number[],
        public readonly base: number = 10
    ) {
        super(domain, range);

        const log = LogScale.logMethodByBase(base);
        const pow = LogScale.powMethodByBase(base);

        this.log = (n: number) => (this.domain[0] >= 0 ? log(n) : -log(-n));
        this.pow = (n: number) => (this.domain[0] >= 0 ? pow(n) : -pow(-n));
    }

    override convert(value: number, clamp?: boolean): number {
        return convertDomain(this.log(value), this.domain.map(this.log), this.range, clamp);
    }

    override invert(value: number, clamp?: boolean): number {
        return convertDomain(value, this.range, this.domain, clamp);
    }

    private static logMethodByBase(base: number) {
        switch (base) {
            case 10:
                return Math.log10;
            case Math.E:
                return Math.log;
            case 2:
                return Math.log2;
            default:
                const logBase = Math.log(base);
                return (x: number) => Math.log(x) / logBase;
        }
    }

    private static powMethodByBase(base: number) {
        switch (base) {
            case 10:
                return (x: number) => (x >= 0 ? 10 ** x : 1 / 10 ** -x);
            case Math.E:
                return Math.exp;
            default:
                return (x: number) => base ** x;
        }
    }
}
