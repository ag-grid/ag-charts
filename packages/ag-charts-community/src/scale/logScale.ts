import { Logger } from '../util/logger';
import { findRangeExtent } from '../util/number';
import { format } from '../util/numberFormat';
import generateTicks, { createNumericTicks, range } from '../util/ticks';
import { ContinuousScale } from './continuousScale';
import { Invalidating } from './invalidating';

const identity = (x: any) => x;

export class LogScale extends ContinuousScale<number> {
    readonly type = 'log';

    public constructor() {
        super([1, 10], [0, 1]);

        // Handling <1 and crossing 0 cases is tricky, easiest solution is to default to clamping.
        this.defaultClampMode = 'clamped';
    }

    toDomain(d: number): number {
        return d;
    }

    @Invalidating
    base = 10;

    protected override transform(x: any) {
        const start = Math.min(this.domain[0], this.domain[1]);
        return start >= 0 ? Math.log(x) : -Math.log(-x);
    }
    protected override transformInvert(x: any) {
        const start = Math.min(this.domain[0], this.domain[1]);
        return start >= 0 ? Math.exp(x) : -Math.exp(-x);
    }

    protected override refresh(): void {
        if (this.base <= 0) {
            this.base = 0;
            Logger.warnOnce('expecting a finite Number greater than to 0');
        }

        super.refresh();
    }

    update() {
        if (!this.domain || this.domain.length < 2) {
            return;
        }
        this.updateLogFn();
        this.updatePowFn();
        if (this.nice) {
            this.updateNiceDomain();
        }
    }

    private baseLog: (x: number) => number = identity;
    private basePow: (x: number) => number = identity;

    private log = (x: number) => {
        const start = Math.min(this.domain[0], this.domain[1]);

        return start >= 0 ? this.baseLog(x) : -this.baseLog(-x);
    };

    private pow = (x: number) => {
        const start = Math.min(this.domain[0], this.domain[1]);
        return start >= 0 ? this.basePow(x) : -this.basePow(-x);
    };

    private updateLogFn() {
        const { base } = this;
        let log: (x: number) => number;
        if (base === 10) {
            log = Math.log10;
        } else if (base === Math.E) {
            log = Math.log;
        } else if (base === 2) {
            log = Math.log2;
        } else {
            const logBase = Math.log(base);
            log = (x) => Math.log(x) / logBase;
        }
        this.baseLog = log;
    }

    private updatePowFn() {
        const { base } = this;
        let pow: (x: number) => number;
        if (base === 10) {
            pow = LogScale.pow10;
        } else if (base === Math.E) {
            pow = Math.exp;
        } else {
            pow = (x) => Math.pow(base, x);
        }
        this.basePow = pow;
    }

    protected updateNiceDomain() {
        const [d0, d1] = this.domain;

        const roundStart = d0 > d1 ? Math.ceil : Math.floor;
        const roundStop = d1 < d0 ? Math.floor : Math.ceil;

        const n0 = this.pow(roundStart(this.log(d0)));
        const n1 = this.pow(roundStop(this.log(d1)));
        this.niceDomain = [n0, n1];
    }

    static pow10(x: number): number {
        return x >= 0 ? Math.pow(10, x) : 1 / Math.pow(10, -x);
    }

    ticks() {
        const count = this.tickCount ?? 10;
        if (!this.domain || this.domain.length < 2 || count < 1) {
            return [];
        }
        this.refresh();
        const base = this.base;
        const [d0, d1] = this.getDomain();

        const start = Math.min(d0, d1);
        const stop = Math.max(d0, d1);

        let p0 = this.log(start);
        let p1 = this.log(stop);

        if (this.interval) {
            const step = Math.abs(this.interval);
            const absDiff = Math.abs(p1 - p0);
            let ticks = range(p0, p1, Math.min(absDiff, step));
            ticks = createNumericTicks(
                ticks.fractionDigits,
                ticks.map((x) => this.pow(x)).filter((t) => t >= start && t <= stop)
            );

            if (!this.isDenseInterval({ start, stop, interval: step, count: ticks.length })) {
                return ticks;
            }
        }

        const isBaseInteger = base % 1 === 0;
        const isDiffLarge = p1 - p0 >= count;

        if (!isBaseInteger || isDiffLarge) {
            // Returns [10^1, 10^2, 10^3, 10^4, ...]
            let ticks = generateTicks(p0, p1, Math.min(p1 - p0, count));
            ticks = createNumericTicks(
                ticks.fractionDigits,
                ticks.map((x) => this.pow(x))
            );
            return ticks;
        }

        const ticks: number[] = [];
        const isPositive = start > 0;
        p0 = Math.floor(p0) - 1;
        p1 = Math.round(p1) + 1;

        const availableSpacing = findRangeExtent(this.range) / count;
        let lastTickPosition = Infinity;
        for (let p = p0; p <= p1; p++) {
            const nextMagnitudeTickPosition = this.convert(this.pow(p + 1));
            for (let k = 1; k < base; k++) {
                const q = isPositive ? k : base - k + 1;
                const t = this.pow(p) * q;
                const tickPosition = this.convert(t);
                const prevSpacing = Math.abs(lastTickPosition - tickPosition);
                const nextSpacing = Math.abs(tickPosition - nextMagnitudeTickPosition);
                const fits = prevSpacing >= availableSpacing && nextSpacing >= availableSpacing;
                if (t >= start && t <= stop && (k === 1 || fits)) {
                    ticks.push(t);
                    lastTickPosition = tickPosition;
                }
            }
        }
        return ticks;
    }

    tickFormat({
        count,
        ticks,
        specifier,
    }: {
        count?: any;
        ticks?: any[];
        specifier?: string | ((x: number) => string);
    }): (x: number) => string {
        const { base } = this;

        if (specifier == null) {
            specifier = base === 10 ? '.0e' : ',';
        }

        if (typeof specifier === 'string') {
            specifier = format(specifier);
        }

        if (count === Infinity) {
            return specifier;
        }

        if (ticks == null) {
            this.ticks();
        }

        return (d) => {
            return (specifier as (x: number) => string)(d);
        };
    }
}
