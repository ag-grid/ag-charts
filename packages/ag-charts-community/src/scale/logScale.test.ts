import { describe, expect, it, test } from '@jest/globals';

import { LogScale } from './logScale';

describe('LogScale', () => {
    test('ticks', () => {
        {
            const scale = new LogScale();
            scale.domain = [100, 1000000];
            expect(scale.ticks()).toEqual({ ticks: [100, 1000, 10000, 100000, 1000000], fractionDigits: 0 });
            scale.tickCount = 4;
            expect(scale.ticks()).toEqual({ ticks: [100, 1000, 10000, 100000, 1000000], fractionDigits: 0 });
        }

        {
            // const scale = new LogScale();
            // scale.domain = [-100, 10000];
            // expect(scale.ticks()).toEqual([]);
        }

        {
            const scale = new LogScale();
            scale.domain = [-1000, -10];
            expect(scale.ticks()).toEqual({ ticks: [-1000, -300, -100, -30, -10], fractionDigits: 0 });
        }
    });

    describe('should create ticks', () => {
        const CASES = [
            {
                interval: 0,
                domain: [0.1, 10000000],
            },
            {
                interval: 1,
                domain: [0.1, 10000000],
            },
            {
                interval: -1,
                domain: [-10000000, -0.1],
            },
            {
                interval: 1,
                domain: [-10000000, -0.1],
            },
            {
                interval: 2,
                domain: [0.1, 10000000],
            },
            {
                interval: 3,
                domain: [0.1, 10000000],
            },
            {
                interval: 4,
                domain: [0.1, 10000000],
            },
            {
                interval: 5,
                domain: [0.1, 10000000],
            },
            {
                interval: 6,
                domain: [0.1, 10000000],
            },
            {
                interval: 7,
                domain: [0.1, 10000000],
            },
            {
                interval: 10,
                domain: [0.1, 10000000],
            },
            {
                interval: 0.5,
                domain: [0.1, 10000000],
            },
            {
                interval: 0.1,
                domain: [0.1, 10000000],
            },
            {
                interval: 2,
                domain: [-10000000, -0.1],
            },
            {
                interval: -2,
                domain: [-10000000, -0.1],
            },
            {
                interval: 10,
                domain: [-10000000, -0.1],
            },
            {
                interval: -10,
                domain: [-10000000, -0.1],
            },
        ];

        it.each(CASES)(`for interval: $interval domain: $domain case`, ({ interval, domain }) => {
            const scale = new LogScale();

            scale.range = [0, 600];
            scale.domain = domain;
            scale.interval = interval;

            expect(scale.ticks()).toMatchSnapshot();
        });
    });

    test('convert', () => {
        {
            const scale = new LogScale();
            scale.domain = [10, 1000];
            expect(scale.convert(50)).toBe(0.3494850021680094);
        }

        {
            const scale = new LogScale();
            scale.domain = [-1000, -10];
            expect(scale.convert(-50)).toBe(0.6505149978319906);
        }
    });

    test('base', () => {
        const expTicks = {
            ticks: [20.085536923187668, 54.598150033144236, 148.4131591025766, 403.4287934927351],
            fractionDigits: 0,
        };
        const scale = new LogScale();
        scale.domain = [10, 1000];
        expect({ ...scale.ticks() }).not.toEqual({ ...expTicks });
        scale.base = Math.E;
        expect({ ...scale.ticks() }).toEqual({ ...expTicks });
    });

    test('nice', () => {
        {
            const scale = new LogScale();
            scale.domain = [57, 775];
            scale.nice = true;
            scale.update();
            expect(scale.nice).toBe(true);
            expect(scale.niceDomain).toEqual([10, 1000]);
        }

        {
            const scale = new LogScale();
            scale.domain = [Math.E * 1.234, Math.E * 5.783];
            scale.base = Math.E;
            scale.nice = true;
            scale.update();
            const domain = scale.niceDomain;
            expect(Math.log(domain[0])).toEqual(1);
            expect(Math.log(domain[1])).toEqual(3);
        }
    });
});
