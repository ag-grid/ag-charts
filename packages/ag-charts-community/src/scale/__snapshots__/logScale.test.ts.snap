// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create ticks for interval: -1 domain: [-10000000, -0.1] case 1`] = `
[
  -10000000,
  -1000000,
  -100000,
  -10000,
  -1000,
  -100,
  -10,
  -1,
  -0.1,
]
`;

exports[`should create ticks for interval: -2 domain: [-10000000, -0.1] case 1`] = `
[
  -10000000,
  -100000,
  -1000,
  -10,
  -0.1,
]
`;

exports[`should create ticks for interval: -10 domain: [-10000000, -0.1] case 1`] = `
[
  -10000000,
  -0.1,
]
`;

exports[`should create ticks for interval: 0 domain: [0.1, 10000000] case 1`] = `
[
  1,
  100,
  10000,
  1000000,
]
`;

exports[`should create ticks for interval: 0.1 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  0.1258925411794167,
  0.15848931924611132,
  0.199526231496888,
  0.251188643150958,
  0.31622776601683794,
  0.39810717055349726,
  0.5011872336272724,
  0.6309573444801932,
  0.7943282347242815,
  1,
  1.2589254117941673,
  1.5848931924611136,
  1.9952623149688795,
  2.51188643150958,
  3.1622776601683795,
  3.9810717055349722,
  5.011872336272722,
  6.309573444801933,
  7.943282347242816,
  10,
  12.589254117941675,
  15.848931924611133,
  19.952623149688797,
  25.118864315095795,
  31.622776601683793,
  39.810717055349734,
  50.11872336272722,
  63.09573444801933,
  79.43282347242814,
  100,
  125.89254117941675,
  158.48931924611142,
  199.52623149688787,
  251.18864315095797,
  316.2277660168379,
  398.1071705534973,
  501.18723362727246,
  630.957344480193,
  794.3282347242814,
  1000,
  1258.9254117941675,
  1584.893192461114,
  1995.2623149688789,
  2511.88643150958,
  3162.277660168379,
  3981.0717055349733,
  5011.872336272725,
  6309.57344480193,
  7943.282347242814,
  10000,
  12589.254117941662,
  15848.93192461114,
  19952.62314968879,
  25118.864315095823,
  31622.776601683792,
  39810.71705534969,
  50118.72336272725,
  63095.7344480193,
  79432.82347242822,
  100000,
  125892.54117941661,
  158489.3192461114,
  199526.23149688786,
  251188.6431509582,
  316227.7660168379,
  398107.1705534969,
  501187.2336272725,
  630957.344480193,
  794328.2347242822,
  1000000,
  1258925.411794166,
  1584893.1924611141,
  1995262.3149688789,
  2511886.4315095823,
  3162277.6601683795,
  3981071.7055349695,
  5011872.336272725,
  6309573.44480193,
  7943282.347242821,
  10000000,
]
`;

exports[`should create ticks for interval: 0.5 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  0.31622776601683794,
  1,
  3.1622776601683795,
  10,
  31.622776601683793,
  100,
  316.2277660168379,
  1000,
  3162.277660168379,
  10000,
  31622.776601683792,
  100000,
  316227.7660168379,
  1000000,
  3162277.6601683795,
  10000000,
]
`;

exports[`should create ticks for interval: 1 domain: [-10000000, -0.1] case 1`] = `
[
  -10000000,
  -1000000,
  -100000,
  -10000,
  -1000,
  -100,
  -10,
  -1,
  -0.1,
]
`;

exports[`should create ticks for interval: 1 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  1,
  10,
  100,
  1000,
  10000,
  100000,
  1000000,
  10000000,
]
`;

exports[`should create ticks for interval: 2 domain: [-10000000, -0.1] case 1`] = `
[
  -10000000,
  -100000,
  -1000,
  -10,
  -0.1,
]
`;

exports[`should create ticks for interval: 2 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  10,
  1000,
  100000,
  10000000,
]
`;

exports[`should create ticks for interval: 3 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  100,
  100000,
]
`;

exports[`should create ticks for interval: 4 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  1000,
  10000000,
]
`;

exports[`should create ticks for interval: 5 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  10000,
]
`;

exports[`should create ticks for interval: 6 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  100000,
]
`;

exports[`should create ticks for interval: 7 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  1000000,
]
`;

exports[`should create ticks for interval: 10 domain: [-10000000, -0.1] case 1`] = `
[
  -10000000,
  -0.1,
]
`;

exports[`should create ticks for interval: 10 domain: [0.1, 10000000] case 1`] = `
[
  0.1,
  10000000,
]
`;
